File: internal/types/pattern.go
package types

import (
	"fmt"
	"regexp"
	"strings"
)

// compilePattern turns a dialect pattern into an anchored full-match regexp.
//
//	"regex:<raw>" -> the remainder used verbatim
//	otherwise     -> friendly tokens expanded; literal runs regexp-escaped
//
// Tokens: <number> -> [0-9]+ , <slug> -> [a-z0-9]+(?:-[a-z0-9]+)* , <date> -> \d{4}-\d{2}-\d{2}
// An unknown <token> is an error rather than a silently-literal match.
//
// NOTE: <date> is a SHAPE check only — it matches the YYYY-MM-DD digit shape but
// accepts impossible calendar dates like 2026-13-45. For real calendar validation
// use `type: date` (which parses and rejects out-of-range days/months); the two
// intentionally disagree.
//
// The result is always anchored to a full match via the \A(?:…)\z wrapper. For a
// raw "regex:" body we first compile the body standalone so an unbalanced group
// (e.g. "regex:a)b") is rejected as an invalid pattern instead of escaping the
// wrapper and silently producing a mis-anchored regexp.
func compilePattern(pattern string) (*regexp.Regexp, error) {
	var body string
	if rest, ok := strings.CutPrefix(pattern, "regex:"); ok {
		// Validate the raw body in isolation: a structurally-unbalanced body
		// (stray ')' / unclosed '(') would otherwise break out of our own
		// \A(?:…)\z wrapper and defeat the full-match guarantee. Compiling
		// "(?:" + body + ")" standalone surfaces that as an error here.
		if _, err := regexp.Compile("(?:" + rest + ")"); err != nil {
			return nil, err
		}
		body = rest
	} else {
		var b strings.Builder
		i := 0
		for i < len(pattern) {
			if pattern[i] == '<' {
				if end := strings.IndexByte(pattern[i:], '>'); end >= 0 {
					tok := pattern[i : i+end+1]
					switch tok {
					case "<number>":
						b.WriteString("[0-9]+")
					case "<slug>":
						b.WriteString("[a-z0-9]+(?:-[a-z0-9]+)*")
					case "<date>":
						b.WriteString(`\d{4}-\d{2}-\d{2}`)
					default:
						return nil, fmt.Errorf("unknown pattern token %q", tok)
					}
					i += end + 1
					continue
				}
			}
			if pattern[i] == '<' {
				b.WriteString(regexp.QuoteMeta("<")) // unmatched '<' (no '>'): emit literal, advance to avoid stalling
				i++
				continue
			}
			next := strings.IndexByte(pattern[i:], '<')
			if next < 0 {
				b.WriteString(regexp.QuoteMeta(pattern[i:]))
				break
			}
			b.WriteString(regexp.QuoteMeta(pattern[i : i+next]))
			i += next
		}
		body = b.String()
	}
	return regexp.Compile(`\A(?:` + body + `)\z`)
}

File: src/core/types/pattern.ts
// Compiles a dialect pattern into an anchored full-match RegExp. Port of Go
// internal/types/pattern.go.
//
//   "regex:<raw>" -> the remainder used verbatim
//   otherwise     -> friendly tokens expanded; literal runs regexp-escaped
//
// Tokens: <number> -> [0-9]+ , <slug> -> [a-z0-9]+(?:-[a-z0-9]+)* , <date> -> \d{4}-\d{2}-\d{2}
// An unknown <token> throws rather than silently matching literally.
//
// <date> is a SHAPE check only (accepts impossible calendar dates like
// 2026-13-45); use `type: date` for real calendar validation — they disagree by
// design. The result is anchored to a full match. JS `$` (no `m` flag) asserts
// absolute end-of-input — no trailing newline — so ^(?:…)$ is the faithful
// equivalent of Go's \A(?:…)\z. A raw "regex:" body is compiled standalone first
// so an unbalanced group is rejected instead of escaping the wrapper.

// quoteMeta escapes the same metacharacters as Go's regexp.QuoteMeta.
function quoteMeta(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function compilePattern(pattern: string): RegExp {
  let body: string;
  if (pattern.startsWith("regex:")) {
    const rest = pattern.slice("regex:".length);
    // Validate the raw body in isolation so a structurally-unbalanced body
    // can't break out of our own ^(?:…)$ wrapper and defeat full-match.
    new RegExp("(?:" + rest + ")"); // throws on invalid
    body = rest;
  } else {
    let b = "";
    let i = 0;
    while (i < pattern.length) {
      if (pattern[i] === "<") {
        const end = pattern.slice(i).indexOf(">");
        if (end >= 0) {
          const tok = pattern.slice(i, i + end + 1);
          switch (tok) {
            case "<number>":
              b += "[0-9]+";
              break;
            case "<slug>":
              b += "[a-z0-9]+(?:-[a-z0-9]+)*";
              break;
            case "<date>":
              b += "\\d{4}-\\d{2}-\\d{2}";
              break;
            default:
              throw new Error(`unknown pattern token ${JSON.stringify(tok)}`);
          }
          i += end + 1;
          continue;
        }
      }
      if (pattern[i] === "<") {
        b += quoteMeta("<"); // unmatched '<' (no '>'): emit literal, advance
        i++;
        continue;
      }
      const next = pattern.slice(i).indexOf("<");
      if (next < 0) {
        b += quoteMeta(pattern.slice(i));
        break;
      }
      b += quoteMeta(pattern.slice(i, i + next));
      i += next;
    }
    body = b;
  }
  return new RegExp("^(?:" + body + ")$");
}
